ludwig/modules/training_hooks.py

```diff
@@ -0,0 +1,103 @@
+import logging
+from abc import ABC, abstractmethod
+
+import torch
+
+logger = logging.getLogger(__name__)
+
+
+class TrainingHook(ABC):
+    """A base class for training hooks in PyTorch.
+
+    This class provides a template for implementing custom training hooks
+    that can be activated, deactivated, and maintain a handle to the hook.
+
+    Attributes:
+        _hook_handle (Optional[torch.utils.hooks.RemovableHandle]): A handle to the
+            registered forward hook, initially set to None.
+    """
+
+    def __init__(self, **kwargs) -> None:
+        self._hook_handle = None
+
+    @abstractmethod
+    def hook_fn(self, module: torch.nn.Module, inputs: torch.tensor, outputs: torch.tensor) -> torch.tensor:
```

Comment:
Is it `torch.tensor` or `torch.Tensor`? We tend to use `torch.Tensor` throughout the code, maybe the other is an alias?

Reply:
They're aliases of each other to my knowledge, but will update to torch.Tensor for consistency